plugins/download_statistics/server/__init__.py

```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+###############################################################################
+#  Copyright Kitware Inc.
+#
+#  Licensed under the Apache License, Version 2.0 ( the "License" );
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+###############################################################################
+
+
+from girder import events
+from girder.constants import AccessType
+from girder.models.item import Item
+from girder.models.file import File
+
+
+def downloadItemStartEvent(event):
+    # Ony count download as started if offset is 0
+    itemModel = Item()
+    item = itemModel.load(event.info['id'], force=True)
+    firstRequest = 'offset' not in event.info['params'] or event.info['params']['offset'] == '0'
+
+    # Get all files in item to increment their download count
+    files = list(itemModel.childFiles(item=item))
+    for file in files:
+        if firstRequest:
+            downloadFileStart(file['_id'])
+        downloadFileRequest(file['_id'])
+
+
+def downloadItemCompleteEvent(event):
+    itemModel = Item()
+    item = itemModel.load(event.info['id'], force=True)
+
+    # Get all files in item to increment their download count
+    files = list(itemModel.childFiles(item=item))
+    for file in files:
+        downloadFileComplete(file['_id'])
+
+
+def downloadFileStartEvent(event):
+    # Ony count download as started if offset is 0
+    fileModel = File()
+    file = fileModel.load(event.info['id'], force=True)
+    firstRequest = 'offset' not in event.info['params'] or event.info['params']['offset'] == '0'
+
+    if firstRequest:
+        downloadFileStart(file['_id'])
+    downloadFileRequest(file['_id'])
+
+
+def downloadFileCompleteEvent(event):
+    fileModel = File()
+    file = fileModel.load(event.info['id'], force=True)
+    downloadFileComplete(file['_id'])
+
+
+def downloadFileStart(fileId):
+    File().increment(query={'_id': fileId}, field='downloadsStarted', amount=1, multi=False)
```

Comment:
[`increment`'](https://github.com/girder/girder/blob/master/girder/models/model_base.py#L447) is implemented via [MongoDB's `$inc` operator](https://docs.mongodb.com/manual/reference/operator/update/inc/). To access nested fields in Mongo, just use a `.`. So
```
.increment(query={'_id': fileId}, field='downloadStats.started', amount=1, multi=False)
```

As an aside, the `multi` argument is not be necessary, as an `_id` query will only ever match a single document.